File: articles/personality-and-institutions-leadership-bureaucracy-and-social-order/python/analyze_personality_institutions.py
#!/usr/bin/env python3
"""Analyze synthetic personality-institutions data.

This script supports the article "Personality and Institutions: Leadership,
Bureaucracy, and Social Order." It demonstrates institutional-unit summaries,
role-fit metrics, risk metrics, and simple regression-style models using
synthetic data only.
"""

from __future__ import annotations

from pathlib import Path
import json
import numpy as np
import pandas as pd


ROOT = Path(__file__).resolve().parents[1]
DATA_PATH = ROOT / "data" / "synthetic_personality_institutions_bureaucracy.csv"
OUTPUTS = ROOT / "outputs"

TRAITS = [
    "conscientiousness",
    "agreeableness",
    "emotional_stability",
    "openness",
    "dark_trait_pressure",
]

OUTCOMES = [
    "leadership_rating",
    "institutional_performance",
    "institutional_trust",
]


def fit_ols(df: pd.DataFrame, y_col: str, x_cols: list[str]) -> dict[str, float]:
    """Fit a small OLS model with numpy and return coefficients."""
    X = df[x_cols].to_numpy(dtype=float)
    X = np.column_stack([np.ones(len(X)), X])
    y = df[y_col].to_numpy(dtype=float)
    beta, *_ = np.linalg.lstsq(X, y, rcond=None)
    names = ["intercept", *x_cols]
    return {name: float(value) for name, value in zip(names, beta)}


def main() -> None:
    OUTPUTS.mkdir(parents=True, exist_ok=True)

    df = pd.read_csv(DATA_PATH)

    required = {
        "participant_id",
        "institutional_unit",
        "role_type",
        *TRAITS,
        "bureaucratic_fit",
        "discretion_level",
        "accountability_strength",
        "role_clarity",
        *OUTCOMES,
    }
    missing = required.difference(df.columns)
    if missing:
        raise ValueError(f"Missing required columns: {sorted(missing)}")

    df["role_fit_index"] = (
        df["bureaucratic_fit"]
        + df["conscientiousness"]
        + df["accountability_strength"]
        + df["role_clarity"]
    ) / 4.0

    df["institutional_risk_index"] = (
        df["dark_trait_pressure"]
        + df["discretion_level"]
        - df["accountability_strength"]
        - df["role_clarity"]
    )

    df["stewardship_index"] = (
        df["conscientiousness"]
        + df["agreeableness"]
        + df["emotional_stability"]
        + df["bureaucratic_fit"]
        + df["accountability_strength"]
        - df["dark_trait_pressure"]
    ) / 5.0

    unit_summary = (
        df.groupby("institutional_unit")
        .agg(
            n=("participant_id", "count"),
            role_fit_mean=("role_fit_index", "mean"),
            risk_mean=("institutional_risk_index", "mean"),
            stewardship_mean=("stewardship_index", "mean"),
            performance_mean=("institutional_performance", "mean"),
            trust_mean=("institutional_trust", "mean"),
        )
        .reset_index()
    )

    role_summary = (
        df.groupby("role_type")
        .agg(
            n=("participant_id", "count"),
            discretion_mean=("discretion_level", "mean"),
            accountability_mean=("accountability_strength", "mean"),
            leadership_mean=("leadership_rating", "mean"),
            performance_mean=("institutional_performance", "mean"),
            trust_mean=("institutional_trust", "mean"),
            risk_mean=("institutional_risk_index", "mean"),
        )
        .reset_index()
    )

    unit_summary.to_csv(OUTPUTS / "python_institutional_unit_summary.csv", index=False)
    role_summary.to_csv(OUTPUTS / "python_role_type_summary.csv", index=False)

    corr_cols = TRAITS + [
        "bureaucratic_fit",
        "discretion_level",
        "accountability_strength",
        "role_clarity",
        "role_fit_index",
        "institutional_risk_index",
        "stewardship_index",
    ] + OUTCOMES

    df[corr_cols].corr().to_csv(OUTPUTS / "python_trait_institution_correlations.csv")

    model_inputs = [
        "conscientiousness",
        "agreeableness",
        "emotional_stability",
        "openness",
        "dark_trait_pressure",
        "bureaucratic_fit",
        "discretion_level",
        "accountability_strength",
        "role_clarity",
    ]

    models = {
        "leadership_rating": fit_ols(df, "leadership_rating", model_inputs),
        "institutional_performance": fit_ols(df, "institutional_performance", model_inputs),
        "institutional_trust": fit_ols(df, "institutional_trust", model_inputs),
        "note": "Synthetic demonstration only; not a personnel-assessment model.",
    }

    with open(OUTPUTS / "python_institutional_models.json", "w", encoding="utf-8") as f:
        json.dump(models, f, indent=2)

    df.to_csv(OUTPUTS / "python_scored_personality_institutions.csv", index=False)

    print("Wrote Python outputs to:", OUTPUTS)


if __name__ == "__main__":
    main()

File: articles/personality-and-institutions-leadership-bureaucracy-and-social-order/r/analyze_personality_institutions.R
# Personality and Institutions: Leadership, Bureaucracy, and Social Order
# R workflow for institutional-personality analysis

root <- normalizePath(file.path(dirname(sys.frame(1)$ofile), ".."))
data_path <- file.path(root, "data", "synthetic_personality_institutions_bureaucracy.csv")
outputs <- file.path(root, "outputs")
dir.create(outputs, showWarnings = FALSE, recursive = TRUE)

data <- read.csv(data_path, stringsAsFactors = FALSE)

required <- c(
  "participant_id",
  "institutional_unit",
  "role_type",
  "conscientiousness",
  "agreeableness",
  "emotional_stability",
  "openness",
  "dark_trait_pressure",
  "bureaucratic_fit",
  "discretion_level",
  "accountability_strength",
  "leadership_rating",
  "institutional_performance",
  "institutional_trust",
  "role_clarity"
)

missing <- setdiff(required, names(data))
if (length(missing) > 0) {
  stop(paste("Missing required columns:", paste(missing, collapse = ", ")))
}

data$role_fit_index <- (
  data$bureaucratic_fit +
  data$conscientiousness +
  data$accountability_strength +
  data$role_clarity
) / 4

data$institutional_risk_index <- (
  data$dark_trait_pressure +
  data$discretion_level -
  data$accountability_strength -
  data$role_clarity
)

data$stewardship_index <- (
  data$conscientiousness +
  data$agreeableness +
  data$emotional_stability +
  data$bureaucratic_fit +
  data$accountability_strength -
  data$dark_trait_pressure
) / 5

unit_summary <- aggregate(
  data[, c("role_fit_index", "institutional_risk_index", "stewardship_index", "institutional_performance", "institutional_trust")],
  by = list(institutional_unit = data$institutional_unit),
  FUN = mean
)

unit_n <- aggregate(
  data$participant_id,
  by = list(institutional_unit = data$institutional_unit),
  FUN = length
)
names(unit_n)[2] <- "n"
unit_summary <- merge(unit_n, unit_summary, by = "institutional_unit")

write.csv(unit_summary, file.path(outputs, "r_institutional_unit_summary.csv"), row.names = FALSE)

corr_cols <- c(
  "conscientiousness",
  "agreeableness",
  "emotional_stability",
  "openness",
  "dark_trait_pressure",
  "bureaucratic_fit",
  "discretion_level",
  "accountability_strength",
  "role_clarity",
  "role_fit_index",
  "institutional_risk_index",
  "stewardship_index",
  "leadership_rating",
  "institutional_performance",
  "institutional_trust"
)

cor_matrix <- cor(data[, corr_cols], use = "pairwise.complete.obs")
write.csv(cor_matrix, file.path(outputs, "r_trait_institution_correlations.csv"))

model_lead <- lm(
  leadership_rating ~ conscientiousness + agreeableness + emotional_stability +
    openness + dark_trait_pressure + bureaucratic_fit + discretion_level +
    accountability_strength + role_clarity,
  data = data
)

model_perf <- lm(
  institutional_performance ~ conscientiousness + agreeableness + emotional_stability +
    openness + dark_trait_pressure + bureaucratic_fit + discretion_level +
    accountability_strength + role_clarity,
  data = data
)

model_trust <- lm(
  institutional_trust ~ conscientiousness + agreeableness + emotional_stability +
    openness + dark_trait_pressure + bureaucratic_fit + discretion_level *
    accountability_strength + role_clarity,
  data = data
)

capture.output(summary(model_lead), file = file.path(outputs, "r_leadership_model_summary.txt"))
capture.output(summary(model_perf), file = file.path(outputs, "r_performance_model_summary.txt"))
capture.output(summary(model_trust), file = file.path(outputs, "r_trust_model_summary.txt"))

write.csv(data, file.path(outputs, "r_scored_personality_institutions.csv"), row.names = FALSE)

cat("Wrote R outputs to:", outputs, "\n")
